File: 22.generate-parentheses.java
/*
 * @lc app=leetcode id=22 lang=java
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> list = new ArrayList<String>();
        char[] buf = new char[n*2];

        helper(n, n, 0, buf, list);

        return list;
    }

    public void helper(int front, int back, int index, char[] buf, List<String> list){
        if(index == buf.length){
            list.add(new String(buf));
            return;
        }

        if(front != 0){
            buf[index] = '(';
            helper(front-1, back, index+1, buf, list);
        }

        if(back > front){
            buf[index] = ')';
            helper(front, back-1, index+1, buf, list);
        }
    }
}
// @lc code=end

File: cpp/22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        std::vector<std::string> res;
        std::string builder;

        build(res, builder, n, 0, 0);

        return res;
    }

    void build(std::vector<std::string>& res, std::string& builder, int n, int open, int closed) {
        if(closed == n) {
            res.push_back(builder);
            return;
        }

        if(open < n) {
            builder.push_back('(');
            build(res, builder, n, open + 1, closed);
            builder.pop_back();
        }

        if(closed < open) {
            builder.push_back(')');
            build(res, builder, n, open, closed + 1);
            builder.pop_back();
        }
    }

};
// @lc code=end

